src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.cs

```diff
@@ -82,12 +82,34 @@ private static bool IsEscapable(SymbolDisplayPartKind kind)
             }
         }
 
-        private static string EscapeIdentifier(string identifier)
+        private static string EscapeIdentifier(string identifier, ISymbol symbol)
         {
-            var kind = SyntaxFacts.GetKeywordKind(identifier);
-            return kind == SyntaxKind.None
-                ? identifier
-                : $"@{identifier}";
+            SyntaxKind kind = SyntaxFacts.GetKeywordKind(identifier);
+            bool escape;
+            switch (kind)
+            {
+                case SyntaxKind.None:
+                    escape = false;
+                    break;
+
+                case SyntaxKind.ThisKeyword:
+                    var parameter = symbol as IParameterSymbol;
+                    if (parameter is null)
+                    {
+                        escape = true;
+                    }
+                    else
+                    {
+                        escape = !parameter.IsThis;
+                    }
+                    break;
+
+                default:
+                    escape = true;
+                    break;
+            }
+
+            return escape ? $"@{identifier}" : identifier;
```

Comment:
couldn't this be written: 

```c#
var kind = SyntaxFacts.GetKeywordKind(identifier);
return kind == SyntaxKind.None || (symbol as IParameterSymbol)?.IsThis == true
               ? identifier
               : $"@{identifier}";
```

All the flow control above seems overly confusing to me.

?

Reply:
Layering prevents this. The extension is in Workspaces, but this code is in compiler.

---
In reply to: [161929932](https://github.com/dotnet/roslyn/pull/23970#discussion_r161929932) [](ancestors = 161929932)